fix: remove ownership key not found exception

NoelStephensUnity opened the pull request:
When removing ownership from an object which is already owned by the server MLAPI throws a KeyNotFound Exception.
[MTT-1607](https://jira.unity3d.com/browse/MTT-1607)
Based on [Github Issue #552 ](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/issues/552)

### PR Checklist
- [X] Have you added a backport label (if needed)? For example, the `type:backport-release-*` label. After you backport the PR, the label changes to `stat:backported-release-*`.
- [X] Have you updated the changelog? Each package has a `CHANGELOG.md` file.

## Changelog
### com.unity.netcode.gameobjects
- Fixed KeyNotFound exception when removing ownership of a newly spawned NetworkObject that is already owned by the server.

## Testing and Documentation

* Includes integration tests.
* No documentation changes or additions were necessary.

NoelStephensUnity committed:
fix

Fix for Github issue #552 where removing ownership immediately after spawning a NetworkObject does not throw exception.

NoelStephensUnity committed:
test

verifies this fix works.

NoelStephensUnity committed:
updated changelog

Updated change log with the fix notes.

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -187,28 +187,50 @@ internal void RemoveOwnership(NetworkObject networkObject)
+                if (NetworkLog.CurrentLogLevel >= LogLevel.Normal)
                 {
-                    NetworkManager.ConnectedClients[networkObject.OwnerClientId].OwnedObjects.RemoveAt(i);
+                    NetworkLog.LogWarning($"Remove ownership ignored:  {networkObject.name} is already set to the default server ownership!");
```
I wonder if this is actually necessary, couldn't we just call this idempotent? "You were trying to remove ownership. It was already owned by server. Nothing to do here"
It seems to me that this warning would just be annoying to devs, because they would have to add unnecessary checks around their code to prevent it

NoelStephensUnity replied:
Yeah, I originally made it silently return but then figured if you are trying to remove ownership immediately after you spawn something (i.e. the next line of code after spawning) then you would get this exception.  This was how the bug was found, and as such I opted to leave the message there as it only will display that message if you spawn a NetworkObject on the server and then immediately try to remove ownership...so if a user does this they know they don't need to remove any ownership at that point since it is already spawned with the default server ownership.

Good call out as this was what I deliberated over the most.  Even though it won't spam (unless a user is trying to do something they don't have to/need to), does that change your thoughts on logging this warning?

becksebenius-unity replied:
Not really - I mean, what are we warning the user about? That they didn't need to call the function? Every piece of information we tell the user has a cost, and since the consequence here is entirely innocuous, I still think it's not in the users best interest to throw this up because there's not really an actual problem here and we've created noise in their logs. Unless I'm misunderstanding here.

I also think there may be additional cases beyond what you described. What about users that are cleaning up an object's state (say, from a disconnected client or a client that no longer is allowed ownership of this object by the game mechanics) and the user's code defensively calls this multiple times? Maybe they don't _need_ to, but there's no actual consequence to them doing so.

That said, I would leave the decision in your hands. It's not a big sticking point for me.

0xFA11 replied:
what about `if (NetworkLog.CurrentLogLevel == LogLevel.Developer)`? :)
(I think `NetworkLog.CurrentLogLevel >= LogLevel.Normal` is a bug, you probably wanted to make it `<=` in the first place)

0xFA11 replied:
but FWIW, +1 on @becksebenius-unity's point, hence why I suggested a middleground with lesser important LogLevel.

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -187,28 +187,50 @@ internal void RemoveOwnership(NetworkObject networkObject)
-
-            var message = new ChangeOwnershipMessage
+            // Make sure the connected client entry exists before trying to remove ownership.
+            if (NetworkManager.ConnectedClients.ContainsKey(networkObject.OwnerClientId))
```
Seems like this could be a `TryGetValue(networkObject.OwnerClientId, out var client)` so that you don't have to index the connected clients lookup multiple times within the block - both for performance and readability sake

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -187,28 +187,50 @@ internal void RemoveOwnership(NetworkObject networkObject)
+            else
             {
-                NetworkManager.NetworkMetrics.TrackOwnershipChangeSent(client.Key, networkObject, size);
+                if (NetworkLog.CurrentLogLevel >= LogLevel.Normal)
```
here again, this should probably be `<=`

NoelStephensUnity committed:
refactor

Using the existing TryGetNetworkClient per Beck's feedback.

NoelStephensUnity committed:
fix

Removing the console log output for when a user tries to remove ownership from a NetworkObject that already is owned by the server.

NoelStephensUnity committed:
fix

Reversing the log error warning level logic per Fatih's feedback.

NoelStephensUnity committed:
Merge branch 'develop' into fix/removeownership-keynotfoundexception

NoelStephensUnity committed:
Merge branch 'develop' into fix/removeownership-keynotfoundexception
